### PR TITLE
Fix2623and3155

### LINKED ISSUE #2623: SmartCard-HSM occasionally fails to generate secp521r1 signature
### Problem Description

Signing with an EC key on secp521r1 fails occasionally with CKR_DATA_INVALID.
See https://github.com/OpenSC/OpenSC/issues/2421

Method used in https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-sc-hsm.c#L1051-L1055 does not work, as the case of datalen == 137 could be valid for both, 512 bit and 521 bit curves. 

### Proposed Resolution

Implement what was proposed by @dengert : https://github.com/OpenSC/OpenSC/issues/2421#issuecomment-1170049939

### PR DESCRIPTION
**Pull request by CardContact**
Fixes #2623 and #3155

Tested with SmartCard-HSM.

##### Checklist

- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [X] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

File: src/libopensc/pkcs15-sc-hsm.c [hunk 1]
```diff
@@ -1076,6 +1076,11 @@ static int sc_pkcs15emu_sc_hsm_add_prkd(sc_pkcs15_card_t * p15card, u8 keyid) {
 	size_t len;
 	int r;
 
+	if (keyid == 0) {
+		// Device authentication key doe not have PKCS#15 meta data
```
Comment (Jakuje) [inline src/libopensc/pkcs15-sc-hsm.c:1080]:
```suggestion
		// Device authentication key does not have PKCS#15 meta data
```

File: src/libopensc/pkcs15-sc-hsm.c [hunk 2]
```diff
@@ -1106,6 +1111,10 @@ static int sc_pkcs15emu_sc_hsm_add_prkd(sc_pkcs15_card_t * p15card, u8 keyid) {
 	if (prkd.type == SC_PKCS15_TYPE_PRKEY_RSA) {
 		r = sc_pkcs15emu_add_rsa_prkey(p15card, &prkd, key_info);
 	} else {
+		if (key_info->field_length == 528) {
+			// Fix a bug for secp521 key generated with OpenSCDP
```
Comment (Jakuje) [inline src/libopensc/pkcs15-sc-hsm.c:1115]: `OpenSCDP`?

**Review by Jakuje — COMMENTED** [review]
just typos to fix. Code wise looks good.

I think this will fix the issue I have with the cards in https://gitlab.com/redhat-crypto/OpenSC/-/pipelines/1303400877, which occasionally fails to sign with the P-521 key.

File: src/libopensc/pkcs15-sc-hsm.c (hunk 2 above)
Comment (Jakuje) [reply src/libopensc/pkcs15-sc-hsm.c:1115]: Thanks! Did not know about that project!

File: src/libopensc/opensc.h [hunk 3]
```diff
@@ -248,6 +248,7 @@ typedef struct sc_security_env {
 	unsigned long flags;
 	int operation;
 	unsigned long algorithm, algorithm_flags;
+	size_t key_size;
```
Comment (frankmorgner) [inline src/libopensc/opensc.h:251]: please make a comment that this is in bits rather than bytes

**Review by frankmorgner — CHANGES_REQUESTED** [review]

File: src/libopensc/opensc.h (hunk 3 above)
Comment (CardContact) [reply src/libopensc/opensc.h]: Changed the field name to key_size_bits.

File: src/libopensc/pkcs15-sc-hsm.c (hunk 1 above)
Comment (CardContact) [reply src/libopensc/pkcs15-sc-hsm.c]: Done

**Review by Jakuje — APPROVED** [review]

**Review by frankmorgner — APPROVED** [review]
Thank you

**Pull request merged**